Add UBI policy parameters to OG-USA code

rickecon opened the pull request:
This PR adds universal basic income (UBI) policy parameters to OG-USA. The theory for this code update is already part of PR #683.

Needs to:
- [x] Pass all current steady-state tests (for UBI = 0 default)
- [x] Pass all current time path and other tests (for UBI = 0 default)
- [x] Write and pass new tests for functions with UBI including UBI > 0 cases
- [x] Test the model with a policy relevant case
* Finish calibration in `OG-USA-Calibration` of within-household demographic composition. This will be in a different PR.

rickecon commented:
OK. This is progress. I think I have the steady-state main code done. Now I have to update the tests. Then I will move to the TPI. However, much of what I did in the SS solution will apply to the TPI solution.

jdebacker commented on ogusa/aggregates.py:
```diff
@@ -303,10 +307,14 @@ def revenue(r, w, b, n, bq, c, Y, L, K, factor, theta, etr_params,
             np.squeeze(p.lambdas) *
             np.tile(np.reshape(p.omega[:p.T, :], (p.T, p.S, 1)),
                     (1, 1, p.J)))
+        pop_weights_ubi = (
```
@rickecon Why not use the `pop_weights` array above?  `pop_weights_ubi` seems to be the same, just a different shape.  Probably better to have the `ubi` array conform to the same dimensions used for other TxSxJ arrays.

rickecon replied:
Ah, this was the easiest way for me to reshape the `pop_weights` array. The `pop_weights` and `pop_weights_ubi` contain the same information, but they are different dimensions. `pop_weights` is dimension (T, S, J). `pop_weights_ubi` is dimension (S, J, T) to be conformable with the `ubi` array. You see in line 317 that I have to multiply these two 3D arrays before summing over the first two dimensions. @jdebacker

jdebacker replied:
Right, but my question is - why is the `ubi` array SxJxT, whenI think all other arrays are TxSxJ?  Seems like breaking from the norm can cause confusion (and duplicating objects as done here).

rickecon replied:
OK. I'll change the household `ubi` array to be (T, S, J).

rickecon commented:
The last commit fixes missing UBI arguments in the `SS.py` code in the `euler_equation_solver()` function. I also adjusted the requisite tests in `test_SS.py`. There are no longer any failing `test_SS.py` tests in the GitHub CI. I ran all the tests in `test_SS.py` locally including the `local` marked tests and this passes all of them. [The whole set of `test_SS.py` tests took about 25 minutes to run.]
```
ogusa/tests/test_SS.py .............................                     [100%]
...
================= 29 passed, 6 warnings in 1976.49s (0:32:56) ==================
```

I haven't test run the model with nonzero UBI yet, and I have not written the new corresponding tests (which I will probably mark `local`). But this is a start.

rickecon commented:
@jdebacker. The last commit updates the dimensions of the household UBI time series array to (T x S x J) as requested in your [comments above](https://github.com/PSLmodels/OG-USA/pull/715#discussion_r598922687) regarding `aggregates.py`. The full suite of tests in `test_SS.py` still pass on my local machine.
```
ogusa/tests/test_SS.py .............................                     [100%]
================= 29 passed, 6 warnings in 1998.07s (0:33:18) ==================
```
And none of the test failures in the GitHub CI are coming from `test_SS.py`.

rickecon commented:
I just ran the steady-state locally with positive small UBI amounts (`run_ubi.py`), and it ran great. The reform was the following, where the amounts represent annual payments per individual in that age category:
```python
ubi_nom_017 = 1000
ubi_nom_1820 = 500
ubi_nom_2164 = 1500
ubi_nom_65p = 500
```
The steady-state finished computing in 10.5 seconds with all equations satisfied and no violations.
```
GE loop errors =  4.163336342344337e-16 [-9.00321484e-16 -7.70217223e-16 -5.89805982e-16 -3.01841885e-16
 -2.94902991e-16 -2.67147415e-16 -5.29090660e-17] -1.5265566588595902e-16
Iteration: 01  Distance:  4.070632627634941e-14
Foreign debt holdings =  0.24915256802280142
Foreign capital holdings =  0.05260133532040023
resource constraint:  -2.3111720870439e-14
Checking constraints on capital, labor, and consumption.
	There were no violations of the constraints on labor  supply.
	There were no violations of the constraints on  consumption.
Maximum error in labor FOC =  1.0791367799356522e-13
Maximum error in savings FOC =  5.062616992290714e-14
It took 10.522576093673706 seconds to get that part done.
run time =  10.522624015808105
```

rickecon commented:
I think the following results show that the UBI policy is now correctly specified in this PR. I Ran the following `run_ubi.py` code locally. Baseline SS and TPI (all UBI parameters set to zero) ran in 37 min 24.6 sec with all the steady-state and TPI solution conditions satisfied and TPI convergence in 21 iterations. Reform SS and TPI (UBI parameters > 0) ran in 40 min 57.9 sec with all the steady-state and TPI solution conditions satisfied and TPI convergence in 21 iterations.

```python
'''
Example script for setting policy and running OG-USA with UBI.
'''

# import modules
import multiprocessing
from distributed import Client
import time
import numpy as np
import os
import taxcalc
from taxcalc import Calculator
from ogusa import output_tables as ot
from ogusa import output_plots as op
from ogusa.execute import runner
from ogusa.constants import REFORM_DIR, BASELINE_DIR
from ogusa.utils import safe_read_pickle


def main():
    # Define parameters to use for multiprocessing
    client = Client()
    num_workers = min(multiprocessing.cpu_count(), 7)
    print('Number of workers = ', num_workers)
    run_start_time = time.time()

    # Grab a reform JSON file already in Tax-Calculator
    # In this example the 'reform' is a change to 2017 law (the
    # baseline policy is tax law in 2018)
    reform_url = ('https://raw.githubusercontent.com/'
                  'PSLmodels/Tax-Calculator/' + taxcalc.__version__ +
                  '/taxcalc/reforms/2017_law.json')
    ref = Calculator.read_json_param_objects(reform_url, None)
    iit_reform = ref['policy']

    # Directories to save data
    CUR_DIR = os.path.dirname(os.path.realpath(__file__))
    base_dir = os.path.join(CUR_DIR, BASELINE_DIR)
    reform_dir = os.path.join(CUR_DIR, REFORM_DIR)

    # Set some OG model parameters
    # See default_parameters.json for more description of these parameters
    alpha_T = np.zeros(50)  # Adjusting the path of transfer spending
    alpha_T[0:2] = 0.09
    alpha_T[2:10] = 0.09 + 0.01
    alpha_T[10:40] = 0.09 - 0.01
    alpha_T[40:] = 0.09
    alpha_G = np.zeros(7)  # Adjusting the path of non-transfer spending
    alpha_G[0:3] = 0.05 - 0.01
    alpha_G[3:6] = 0.05 - 0.005
    alpha_G[6:] = 0.05
    # Set UBI parameters
    ubi_nom_017 = 1000
    ubi_nom_1820 = 500
    ubi_nom_2164 = 1500
    ubi_nom_65p = 500
    # Set start year for baseline and reform.
    START_YEAR = 2021
    # Also adjust the Frisch elasticity, the start year, the
    # effective corporate income tax rate, and the SS debt-to-GDP ratio
    og_spec_base = {'frisch': 0.41, 'start_year': START_YEAR,
                    'cit_rate': [0.21], 'debt_ratio_ss': 1.0,
                    'alpha_T': alpha_T.tolist(), 'alpha_G': alpha_G.tolist()}

    '''
    ------------------------------------------------------------------------
    Run baseline policy first
    ------------------------------------------------------------------------
    '''
    tax_func_path = os.path.join(
        CUR_DIR, '..', 'ogusa', 'data', 'tax_functions',
        'TxFuncEst_baseline_CPS.pkl')  # use cached baseline estimates
    kwargs = {'output_base': base_dir, 'baseline_dir': base_dir,
              'test': False, 'time_path': True, 'baseline': True,
              'og_spec': og_spec_base, 'guid': '_ubi',
              'run_micro': False, 'tax_func_path': tax_func_path,
              'data': 'cps', 'client': client, 'num_workers': num_workers}

    start_time = time.time()
    runner(**kwargs)
    print('run time = ', time.time()-start_time)

    '''
    ------------------------------------------------------------------------
    Run reform policy
    ------------------------------------------------------------------------
    '''
    # update the effective corporate income tax rate
    og_spec_ref = {'frisch': 0.41, 'start_year': START_YEAR,
                   'cit_rate': [0.35], 'debt_ratio_ss': 1.0,
                   'alpha_T': alpha_T.tolist(), 'alpha_G': alpha_G.tolist(),
                   'ubi_nom_017': ubi_nom_017, 'ubi_nom_1820': ubi_nom_1820,
                   'ubi_nom_2164': ubi_nom_2164, 'ubi_nom_65p': ubi_nom_65p}
    kwargs = {'output_base': reform_dir, 'baseline_dir': base_dir,
              'test': False, 'time_path': True, 'baseline': False,
              'og_spec': og_spec_ref, 'guid': '_ubi', 'run_micro': False,
              'tax_func_path': tax_func_path, 'data': 'cps',
              'client': client, 'num_workers': num_workers}

    start_time = time.time()
    runner(**kwargs)
    print('run time = ', time.time()-start_time)

    # return ans - the percentage changes in macro aggregates and prices
    # due to policy changes from the baseline to the reform
    base_tpi = safe_read_pickle(
        os.path.join(base_dir, 'TPI', 'TPI_vars.pkl'))
    base_params = safe_read_pickle(
        os.path.join(base_dir, 'model_params.pkl'))
    reform_tpi = safe_read_pickle(
        os.path.join(reform_dir, 'TPI', 'TPI_vars.pkl'))
    reform_params = safe_read_pickle(
        os.path.join(reform_dir, 'model_params.pkl'))
    ans = ot.macro_table(
        base_tpi, base_params, reform_tpi=reform_tpi,
        reform_params=reform_params,
        var_list=['Y', 'C', 'K', 'L', 'r', 'w'], output_type='pct_diff',
        num_years=10, start_year=og_spec_ref['start_year'])

    # create plots of output
    op.plot_all(base_dir, reform_dir,
                os.path.join(CUR_DIR, 'run_example_plots'))

    print("total time was ", (time.time() - run_start_time))
    print('Percentage changes in aggregates:', ans)
    # save percentage change output to csv file
    ans.to_csv('ogusa_example_output.csv')
    client.close()


if __name__ == "__main__":
    # execute only if run as a script
    main()
```

The **baseline steady-state** terminal output was the following.
```
GE loop errors =  -7.834011217511261e-14 [1.29063427e-15 1.75207071e-15 2.70616862e-15 3.00801051e-15
 3.03403136e-15 3.25087179e-15 9.49327422e-16] 1.1012024625500771e-14 -2.581268532253489e-15
Iteration: 01  Distance:  8.750963943201238e-13
Foreign debt holdings =  0.25011849591737656
Foreign capital holdings =  0.05620836283554873
resource constraint:  -1.1556901269305087e-13
Checking constraints on capital, labor, and consumption.
	There were no violations of the constraints on labor  supply.
	There were no violations of the constraints on  consumption.
Maximum error in labor FOC =  1.0347278589506459e-13
Maximum error in savings FOC =  5.639932965095795e-14
```

The **baseline transition path** terminal output was the following.
```
Type  4  max euler error =  1.092459456231154e-12
Type  1  max euler error =  2.63339350325964e-12
Type  6  max euler error =  8.215650382226158e-13
Type  5  max euler error =  5.924150059399835e-13
Type  2  max euler error =  1.998595733354591e-12
Type  0  max euler error =  3.4495184486615926e-12
Type  3  max euler error =  1.7527090889757346e-12
Maximum debt ratio:  1.623080152732493
r diff:  2.4050836516797958e-08 -2.82720107741552e-07
BQ diff:  1.2026442222645706e-07 -2.7859878867428356e-08
TR diff:  2.6202417457987792e-09 -2.751556954644352e-07
Y diff:  2.911379726189267e-08 -3.439446193298501e-06
Iteration: 21
	Distance: 6.751348325281791e-06
Max absolute value resource constraint error: 3.679064008596522e-06
Checking time path for violations of constraints.
...
Time path iteration complete.
It took 2244.5693321228027 seconds to get that part done.
run time =  2244.5761001110077
```

The **reform steady-state** terminal output was the following.
```
GE loop errors =  4.163336342344337e-16 [-9.00321484e-16 -7.70217223e-16 -5.89805982e-16 -3.01841885e-16
 -2.94902991e-16 -2.67147415e-16 -5.29090660e-17] -1.5265566588595902e-16
Iteration: 01  Distance:  4.070632627634941e-14
Foreign debt holdings =  0.24915256802280142
Foreign capital holdings =  0.05260133532040023
resource constraint:  -2.3111720870439e-14
Checking constraints on capital, labor, and consumption.
	There were no violations of the constraints on labor  supply.
	There were no violations of the constraints on  consumption.
Maximum error in labor FOC =  1.0791367799356522e-13
Maximum error in savings FOC =  5.062616992290714e-14
```

The **reform transition path** terminal output was the following.
```
Type  4  max euler error =  1.1501910535116622e-12
Type  1  max euler error =  2.425032397113114e-12
Type  2  max euler error =  1.8302026560945706e-12
Type  6  max euler error =  9.85433956657289e-13
Type  3  max euler error =  1.4842294060457561e-12
Type  5  max euler error =  6.126210649881614e-13
Type  0  max euler error =  4.524602914557363e-12
Maximum debt ratio:  2.1359154241947524
r diff:  2.441155957932928e-09 -2.410347988146544e-07
BQ diff:  1.1792340940883217e-07 -2.3433122491034464e-08
TR diff:  2.1377680375911368e-08 -2.8758497070541233e-07
Y diff:  2.1377680370360252e-07 -3.594812133855818e-06
Iteration: 21
	Distance: 6.589222826270178e-06
Max absolute value resource constraint error: 3.716198857323799e-06
Checking time path for violations of constraints.
...
Time path iteration complete.
It took 2457.9075379371643 seconds to get that part done.
run time =  2457.914320230484
Percentage changes in aggregates: Year                    Variable  2021  2022  2023  2024  2025  2026  2027  2028  2029  2030  2021-2030    SS
0                    GDP ($Y_t$) -0.92 -1.00 -1.19 -1.31 -1.41 -1.50 -1.59 -1.70 -1.80 -1.91      -1.42 -0.39
1            Consumption ($C_t$)  1.74  3.03  2.09  1.79  1.64  1.53  1.44  1.34  1.28  1.22       1.71 -0.72
2          Capital Stock ($K_t$) -0.52 -0.73 -1.21 -1.54 -1.83 -2.11 -2.38 -2.64 -2.92 -3.22      -1.90 -1.51
3                  Labor ($L_t$) -1.13 -1.14 -1.17 -1.18 -1.18 -1.17 -1.17 -1.19 -1.19 -1.20      -1.17  0.22
4     Real interest rate ($r_t$) -4.14 -3.90 -3.38 -3.01 -2.66 -2.33 -2.00 -1.74 -1.37 -1.00      -2.60 -1.40
5                      Wage rate  0.22  0.15 -0.01 -0.13 -0.23 -0.33 -0.43 -0.51 -0.62 -0.72      -0.27 -0.61
```

The transition path for the Debt-to-GDP ratio of baseline and reform is the following:
![DebtGDPratio](https://user-images.githubusercontent.com/4317690/117084343-2fe5dd00-ad04-11eb-8261-86b01c93959e.png)

The percent change in the macro aggregates from the reform across the transition path is the following:
![MacroAgg_PctChange](https://user-images.githubusercontent.com/4317690/117084448-776c6900-ad04-11eb-8efe-3fbf50c7057a.png)

cc: @jdebacker @MaxGhenis

rickecon commented:
@jdebacker @MaxGhenis Now I am working on getting the tests to pass and writing new tests. All the CI tests that are currently failing are simply failing because of modified functions in the code (added UBI) that have not been updated in the tests.
```
=========================== short test summary info ============================
FAILED ogusa/tests/test_TPI.py::test_firstdoughnutring - TypeError: firstdoug...
FAILED ogusa/tests/test_TPI.py::test_twist_doughnut[s<S] - TypeError: twist_d...
FAILED ogusa/tests/test_TPI.py::test_twist_doughnut[s==S] - TypeError: twist_...
FAILED ogusa/tests/test_aggregates.py::test_revenue[SS] - TypeError: revenue(...
FAILED ogusa/tests/test_aggregates.py::test_revenue[TPI] - TypeError: revenue...
FAILED ogusa/tests/test_aggregates.py::test_revenue[TPI, replace rate adjust]
FAILED ogusa/tests/test_fiscal.py::test_D_G_path[baseline_spending = False]
FAILED ogusa/tests/test_fiscal.py::test_D_G_path[baseline_spending = True] - ...
FAILED ogusa/tests/test_fiscal.py::test_D_G_path[balanced_budget = True] - Va...
FAILED ogusa/tests/test_fiscal.py::test_get_G_ss[balanced_budget = False] - T...
FAILED ogusa/tests/test_fiscal.py::test_get_G_ss[balanced_budget = True] - Ty...
FAILED ogusa/tests/test_fiscal.py::test_get_TR[balanced_budget = False] - Typ...
FAILED ogusa/tests/test_fiscal.py::test_get_TR[balanced_budget = True] - Type...
FAILED ogusa/tests/test_fiscal.py::test_get_TR[baseline_spending = True] - Ty...
FAILED ogusa/tests/test_fiscal.py::test_get_TR[Time path, balanced_budget = False]
FAILED ogusa/tests/test_household.py::test_FOC_savings[SS] - TypeError: FOC_s...
FAILED ogusa/tests/test_household.py::test_FOC_savings[TPI] - TypeError: FOC_...
FAILED ogusa/tests/test_household.py::test_FOC_savings[SS - wealth tax] - Typ...
FAILED ogusa/tests/test_household.py::test_FOC_savings[TPI - wealth tax] - Ty...
FAILED ogusa/tests/test_household.py::test_FOC_savings[SS - j =0] - TypeError...
FAILED ogusa/tests/test_household.py::test_FOC_labor[SS] - TypeError: FOC_lab...
FAILED ogusa/tests/test_household.py::test_FOC_labor[TPI] - TypeError: FOC_la...
FAILED ogusa/tests/test_household.py::test_FOC_labor[vary tau_payroll] - Type...
FAILED ogusa/tests/test_tax.py::test_net_taxes[SS] - TypeError: net_taxes() m...
FAILED ogusa/tests/test_tax.py::test_net_taxes[TPI Scalar] - TypeError: net_t...
FAILED ogusa/tests/test_tax.py::test_net_taxes[TPI shift = True] - TypeError:...
FAILED ogusa/tests/test_tax.py::test_net_taxes[TPI shift = False] - TypeError...
FAILED ogusa/tests/test_tax.py::test_net_taxes[TPI 3D] - TypeError: net_taxes...
FAILED ogusa/tests/test_tax.py::test_net_taxes[TPI 3D,vary tau_bq] - TypeErro...
FAILED ogusa/tests/test_tax.py::test_net_taxes[TPI 3D,vary retire] - TypeErro...
FAILED ogusa/tests/test_tax.py::test_net_taxes[TPI 3D,vary replacement rate]
== 31 failed, 368 passed, 401 deselected, 679 warnings in 1746.43s (0:29:06) ===
```

rickecon commented:
@jdebacker. The last two tests that I have to fix are in `test_TPI.py`, and they are the two `test_twist_doughnut()` tests. I need to add an 80 x 7 numpy array of zeros as the seventh element of the two input tuples saved as pickle files `/ogusa/tests/test_io_data/twist_doughnut_inputs_2.pkl` and `/ogusa/tests/test_io_data/twist_doughnut_inputs_2.pkl`,  respectively. In both cases, this will convert the two tuples from each being length 16 to length 17. I am having trouble updating these tuples using the `pickle.dump()` command.  I get the error:
```
Can't pickle <class 'marshmallow.schema.DefaultsSchema'>: attribute lookup DefaultsSchema on marshmallow.schema failed
```
I remember something like this coming from the pickle method we use with Compute.Studio. Can you shed some light on this problem? Thanks.

jdebacker commented on ogusa/tests/test_TPI.py:
```diff
@@ -166,9 +167,9 @@ def test_twist_doughnut(file_inputs, file_outputs):
     '''
     input_tuple = utils.safe_read_pickle(file_inputs)
-    (guesses, r, w, bq, tr, theta, factor, j, s, t, tau_c, etr_params,
+    (guesses, r, w, bq, tr, theta, factor, ubi, j, s, t, tau_c, etr_params,
```
@rickecon You don't need to update the `input_tuple` to include the `ubi` object.  Rather, it'll be easier to read in the `input_tuple` as is done and then add a line below to create the `ubi` object.  e.g., `ubi = np.zeros(int(len(guesses) / 2))`.

jdebacker commented:
@rickecon See my comment above -- I do not think you'll should update this pickle file as you can just create the `ubi` object in the test.

But, FYI, the issue is the the pickle file contains a `Specifications` class object and so would need to be pickled with `cloudpickle`.

rickecon commented:
@jdebacker . This PR is ready for review.

rickecon commented:
@jdebacker . I am just going to run the full battery of tests locally before this PR is ready to be merged. But it is now passing all the CI tests.

rickecon commented:
@jdebacker. I just ran the full `pytest`. It has been running for 12 hours, but I think it has stalled on `test_startyears.py` for the last two hours. I am going to stop it and manually run the remaining test modules. But the results of my local tests that successfully ran are the following:
```
collected 806 items                                                            

ogusa/tests/test_SS.py ..............................                    [  3%]
ogusa/tests/test_TPI.py ......F...............                           [  6%]
ogusa/tests/test_aggregates.py ...............................           [ 10%]
ogusa/tests/test_basic.py ....                                           [ 10%]
ogusa/tests/test_demographics.py ........                                [ 11%]
ogusa/tests/test_elliptical_u_est.py .......                             [ 12%]
ogusa/tests/test_execute.py ..                                           [ 12%]
ogusa/tests/test_firm.py ..........................                      [ 16%]
ogusa/tests/test_fiscal.py ...............                               [ 17%]
ogusa/tests/test_get_micro_data.py ..............                        [ 19%]
ogusa/tests/test_household.py ........................................   [ 24%]
ogusa/tests/test_income.py ............                                  [ 26%]
ogusa/tests/test_output_plots.py .....................................   [ 30%]
ogusa/tests/test_output_tables.py .............                          [ 32%]
ogusa/tests/test_parameter_plots.py ...................................  [ 36%]
ogusa/tests/test_parameter_tables.py .......                             [ 37%]
ogusa/tests/test_parameters.py .................                         [ 39%]
ogusa/tests/test_run_example.py .                                        [ 39%]
ogusa/tests/test_run_ogusa.py .                                          [ 39%]
...
```
The one fail is in `test_TPI.py` in the local run `test_inner_loop()` function. I will report when that is fixed and I will report when the rest of the tests run successfully.

rickecon commented:
@jdebacker. Five more test modules passed.
```
ogusa/tests/test_startyears.py ...                                       [100%]
ogusa/tests/test_tax.py ...............................                  [100%]
ogusa/tests/test_user_inputs.py .........                                [100%]
ogusa/tests/test_utils.py .............................................. [ 71%]
..................                                                       [100%]
ogusa/tests/test_wealth.py ..                                            [100%]
```
There are two remaining errors in `test_txfunc.py` that I need to fix. Once those are done, all tests will have passed and this PR will be ready for review and merge.
```
ogusa/tests/test_txfunc.py .............F..........F                     [100%]
...
=========================== short test summary info ============================
FAILED ogusa/tests/test_txfunc.py::test_tax_func_loop - assert False
FAILED ogusa/tests/test_txfunc.py::test_tax_func_estimate - assert False
============= 2 failed, 23 passed, 1 warning in 568.35s (0:09:28) ==============

```

jdebacker commented:
@rickecon, FYI, the failing `test_txfunc.py::test_tax_func_loop` test (and maybe the other failing test in `test_txfunc.py`)  has nothing to do with your UBI modifications, but rather is related to updated package versions for the OG-USA dependencies.  If PR #725, I updated the output for `test_tax_func_loop` to be consistent with what is now returned.  You'll probably just need to do the same here.

rickecon commented:
@jdebacker. This PR is ready for final review and merge. It now passes all the test, both CI and local run.

Regarding the two failing tests in `test_txfunc.py`, you were right. The issue with the `test_tax_func_loop()` test is clearly an estimation difference, probably stemming from `scipy` version. I checked the `test_tuple` against the `expected_tuple` in that test, and 14 out of 16 of the elements satisfied the `np.allclose()` condition. However, the maximum absolute distance between elements of the 11th element of those two tuples (80 x 12 matrix of tax function parameters by age) was about 0.75. That ended up being a small percent difference. The same was true of the 16th element. The same was true for the `expected_dict` and `test_dict` dictionaries in the `test_tax_func_estimate()` test. If these two tests don't pass on your computer, we should probably just rewrite this test in terms of percent deviations. I also deleted some extraneous code from both of these tests.
```
collected 25 items                                                             

ogusa/tests/test_txfunc.py .........................                     [100%]

================== 25 passed, 1 warning in 627.94s (0:10:27) ===================
```

jdebacker commented on ogusa/tests/test_SS.py:
```diff
@@ -933,6 +935,121 @@ def test_euler_equation_solver(dask_client):
     assert(np.allclose(np.array(test_list), np.array(expected_list)))
 
 
+@pytest.mark.local
```
Why is the a local test? Isn't this a quick test to run?

rickecon replied:
@jdebacker . `UBI>0` is not a baseline use case, and I didn't want to add time to the already close-to-the-limit CI test suite. So my approach was to separate all `UBI>0` tests to be run locally and not in the CI.

jdebacker commented on ogusa/tests/test_tax.py:
```diff
@@ -708,39 +718,163 @@ def test_get_biz_tax():
+              ubi_u[:p_u.T, :, j3], theta3, 0, j3, shift3, method3, p_u.e[:, j3],
+              etr_params3, p_u, expected3)]
+
+@pytest.mark.local
```
Why a `local` marker?  Why not just another parameterization of the `test_net_taxes` function above?

rickecon replied:
@jdebacker see comment above.

jdebacker commented:
@rickecon This is looking really good.  I've got a couple comments on some unit tests.

Also, it'd be helpful to update the API documentation files at OG-USA/docs/book/content/api/ to list any new functions you added to the relevant modules.

rickecon commented:
@jdebacker. I updated all the docstrings of functions for which I either had to add UBI as an input argument or an output argument. This should update automatically when the API documentation compiles in the new repo.
* `SS.py`: `euler_equation_solver()`, 
* `household.py`: `FOC_savings()`, `FOC_labor()`
* `tax.py`: `net_taxes()`
* `aggregates.py`: `revenue()`
* `fiscal.py`: `D_G_path()`, `get_G_ss()`, `get_TR()`, and updated the LaTeX equation in the docstrings
* `TPI.py`: `firstdoughnutring()`, `twist_doughnut()`, `inner_loop()`

The only API function that needs to be added to the API documentation because it is new is `get_ubi_nom_objs()` in the `Specifications` class in `parameters.py`. I just added that, and it should show up in the API documentation upon the first compilation of the documentation after the merge. I updated `./docs/book/content/api/parameters.rst`.

rickecon commented:
@jdebacker . I consolidated all the tests with `UBI>0`, removed the `local` testing markers, and removed the `baseline` labels in the docstrings separating tests. This is ready for your final review.

jdebacker commented:
Excellent contribution @rickecon!  Merging.
